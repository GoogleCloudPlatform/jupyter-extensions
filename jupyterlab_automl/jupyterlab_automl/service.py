"""Services for AutoML extension backend"""

import math
from collections import defaultdict
from enum import Enum

from google.cloud import automl_v1beta1

from gcp_jupyterlab_shared.handlers import AuthProvider


def parse_dataset_type(dataset):
    for dt in DatasetType:
        if dt.value == dataset.name.split("/")[-1][:3]:
            return dt
    return DatasetType.other


class DatasetType(Enum):
    other = "other"
    TBL = "TBL"
    ICN = "ICN"
    IOD = "IOD"


class ColumnType(Enum):
    Unspecified = 0
    Numeric = 3
    Timestamp = 4
    String = 6
    Array = 8
    Struct = 9
    Categorical = 10
    Unrecognized = -1


class Months(Enum):
    Jan = 1
    Feb = 2
    Mar = 3
    Apr = 4
    May = 5
    Jun = 6
    Jul = 7
    Aug = 8
    Sep = 9
    Oct = 10
    Nov = 11
    Dec = 12


class Days(Enum):
    Mon = 1
    Tue = 2
    Wed = 3
    Thu = 4
    Fri = 5
    Sat = 6
    Sun = 7


class ChartInfo(Enum):
    name = "name"
    amount = "Number of Instances"


class AutoMLService:
    """Provides an authenticated AutoML Client and project info"""

    _instance = None

    def __init__(self):
        self._client = automl_v1beta1.AutoMlClient()
        self._parent = self._client.location_path(
            AuthProvider.get().project, "us-central1"
        )

    @property
    def client(self):
        return self._client

    @classmethod
    def get(cls):
        if not cls._instance:
            cls._instance = AutoMLService()
        return cls._instance

    def get_models(self):
        models = self._client.list_models(self._parent)
        return {
            "models": [
                {
                    "id": model.name,
                    "displayName": model.display_name,
                    "datasetId": model.dataset_id,
                    "updateTime": model.update_time.ToMilliseconds(),
                    "deploymentState": model.deployment_state,
                    "metadata": "",
                }
                for model in models
            ]
        }

    def get_datasets(self):
        datasets = []
        for dataset in self._client.list_datasets(self._parent):
            dataset_type, metadata = self._get_dataset_metadata(dataset)
            if dataset_type != DatasetType.other:
                datasets.append(
                    {
                        "id": dataset.name,
                        "displayName": dataset.display_name,
                        "description": dataset.description,
                        "createTime": dataset.create_time.ToMilliseconds(),
                        "exampleCount": dataset.example_count,
                        "datasetType": dataset_type.value,
                        "metadata": metadata,
                    }
                )
        return {"datasets": datasets}

    def _get_dataset_metadata(self, dataset):
        dataset_type = parse_dataset_type(dataset)
        if dataset_type == DatasetType.TBL:
            metadata = {
                "primary_table_spec_id": dataset.tables_dataset_metadata.primary_table_spec_id,
                "target_column_spec_id": dataset.tables_dataset_metadata.target_column_spec_id,
                "weight_column_spec_id": dataset.tables_dataset_metadata.weight_column_spec_id,
                "ml_use_column_spec_id": dataset.tables_dataset_metadata.ml_use_column_spec_id,
                "stats_update_time": dataset.tables_dataset_metadata.stats_update_time.ToMilliseconds(),
            }
        elif dataset_type == DatasetType.ICN:
            metadata = {
                "classification_type": dataset.image_classification_dataset_metadata.classification_type,
            }
        elif dataset_type == DatasetType.IOD:
            metadata = ""
        return dataset_type, metadata

    def get_table_specs(self, dataset_id):
        table_specs = []
        for table_spec in self._client.list_table_specs(dataset_id):
            column_spec, chart_summary = self.get_column_specs(table_spec)
            table_specs.append(
                {
                    "id": table_spec.name,
                    "rowCount": table_spec.row_count,
                    "validRowCount": table_spec.valid_row_count,
                    "columnCount": table_spec.column_count,
                    "columnSpecs": column_spec,
                    "chartSummary": chart_summary,
                }
            )
        return {"tableSpecs": table_specs}

    def get_column_specs(self, table_spec):
        column_specs = []
        type_summary = defaultdict(int)
        for column_spec in self._client.list_column_specs(table_spec.name):
            try:
                type_code = ColumnType(column_spec.data_type.type_code).name
            except:
                type_code = ColumnType.unrecognized.name
            detail_panel = self._get_detail_panel(
                column_spec,
                table_spec.row_count - column_spec.data_stats.null_value_count,
            )
            type_summary[type_code] += 1
            column_specs.append(
                {
                    "id": column_spec.name,
                    "dataType": type_code,
                    "displayName": column_spec.display_name,
                    "distinctValueCount": column_spec.data_stats.distinct_value_count,
                    "invalidValueCount": table_spec.row_count
                    - column_spec.data_stats.valid_value_count,
                    "nullValueCount": str(column_spec.data_stats.null_value_count)
                    + " ("
                    + str(
                        math.floor(
                            column_spec.data_stats.null_value_count
                            / table_spec.row_count
                            * 100
                        )
                    )
                    + "%)",
                    "nullable": column_spec.data_type.nullable,
                    "detailPanel": detail_panel,
                }
            )
        columns_summary = []
        for key, val in type_summary.items():
            columns_summary.append(
                {ChartInfo.name.value: key, ChartInfo.amount.value: val}
            )
        return column_specs, columns_summary

    def _get_bucket_label(self, bucket):
        if bucket.min == float("-inf"):
            return "[" + str(bucket.min) + ", " + str(round(bucket.max)) + "]"
        elif bucket.max == float("inf"):
            return "[" + str(round(bucket.min)) + ", " + str(bucket.max) + "]"
        else:
            return "[" + str(round(bucket.min)) + ", " + str(round(bucket.max)) + "]"

    def _get_detail_panel(self, column_spec, count):
        chart_data = []
        if column_spec.data_type.type_code == 3:
            mean = round(column_spec.data_stats.float64_stats.mean, 2)
            standard_deviation = round(
                column_spec.data_stats.float64_stats.standard_deviation, 2
            )
            for bucket in column_spec.data_stats.float64_stats.histogram_buckets:
                chart_data.append(
                    {
                        ChartInfo.name.value: self._get_bucket_label(bucket),
                        ChartInfo.amount.value: bucket.count,
                    }
                )
            return [chart_data, mean, standard_deviation]
        elif column_spec.data_type.type_code == 10:
            try:
                div = (
                    column_spec.data_stats.category_stats.top_category_stats[0].count
                    / count
                )
                rounded = round(div * 100, 3)
                most_common = (
                    column_spec.data_stats.category_stats.top_category_stats[0].value
                    + " ("
                    + str(rounded)
                    + "%)"
                )
            except:
                most_common = ""
            for stat in column_spec.data_stats.category_stats.top_category_stats:
                chart_data.append(
                    {
                        ChartInfo.name.value: stat.value,
                        ChartInfo.amount.value: stat.count,
                    }
                )
            return [chart_data, most_common]
        elif column_spec.data_type.type_code == 4:
            month_chart = []
            day_chart = []
            time_chart = []
            for month, amount in dict(
                column_spec.data_stats.timestamp_stats.granular_stats[
                    "month_of_year"
                ].buckets
            ).items():
                month_chart.append(
                    {
                        ChartInfo.name.value: Months(month).name,
                        ChartInfo.amount.value: amount,
                    }
                )
            for day, amount in dict(
                column_spec.data_stats.timestamp_stats.granular_stats[
                    "day_of_week"
                ].buckets
            ).items():
                day_chart.append(
                    {
                        ChartInfo.name.value: Days(day).name,
                        ChartInfo.amount.value: amount,
                    }
                )
            for hour, amount in dict(
                column_spec.data_stats.timestamp_stats.granular_stats[
                    "hour_of_day"
                ].buckets
            ).items():
                time_chart.append(
                    {
                        ChartInfo.name.value: str(hour) + ":00",
                        ChartInfo.amount.value: amount,
                    }
                )
            return [month_chart, day_chart, time_chart]
        else:
            return []
