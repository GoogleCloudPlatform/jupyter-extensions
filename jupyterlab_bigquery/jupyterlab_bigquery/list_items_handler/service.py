# Lint as: python3
"""Request handler classes for the extensions."""

import re
from google.cloud import bigquery
from google.cloud.datacatalog import DataCatalogClient, enums, types

SCOPE = ("https://www.googleapis.com/auth/cloud-platform",)


class BigQueryService:
  """Provides an authenticated BigQuery Client and project info"""

  _instance = None

  def __init__(self,
               client=bigquery.Client(),
               datacatalog_client=DataCatalogClient()):
    self._client = client
    self._datacatalog_client = datacatalog_client

  @property
  def client(self):
    return self._client

  @classmethod
  def get(cls):
    if not cls._instance:
      cls._instance = BigQueryService()
    return cls._instance

  def list_projects(self):
    project = self._client.project
    projects_list = {
      format(project): {
        'id': format(project),
        'name': format(project),
      }
    }
    project_ids = [format(project)]
    return {'projects': projects_list, 'projectIds': project_ids}

  def list_datasets(self, project):
    datasets = list(self._client.list_datasets(project))

    datasets_list = {}
    dataset_ids = []
    for dataset in datasets:
      dataset_full_id = '{}.{}'.format(dataset.project, dataset.dataset_id)
      datasets_list[dataset_full_id] = {
          'id': dataset_full_id,
          'name': dataset.dataset_id,
          'projectId': format(dataset.project),
      }
      dataset_ids.append(dataset_full_id)

    return {'datasets': datasets_list, 'datasetIds': dataset_ids}

  def list_tables(self, dataset_id):
    dataset = self._client.get_dataset(dataset_id)
    tables = list(self._client.list_tables(dataset))

    tables_list = {}
    table_ids = []
    for table in tables:
      table_full_id = '{}.{}.{}'.format(table.project, table.dataset_id, table.table_id)
      tables_list[table_full_id] = {
        'id': table_full_id,
        'name': table.table_id,
        'datasetId': dataset_id,
        'type': table.table_type,
        'partitioned': True if table.partitioning_type else False,
      }
      table_ids.append(table_full_id)

    return {'tables': tables_list, 'tableIds': table_ids}

  def list_models(self, dataset_id):
    dataset = self._client.get_dataset(dataset_id)
    models = list(self._client.list_models(dataset))

    models_list = {}
    model_ids = []
    for model in models:
      model_full_id = '{}.{}.{}'.format(model.project, model.dataset_id, model.model_id)
      models_list[model_full_id] = {
        'id': model_full_id,
        'name': model.model_id,
        'datasetId': dataset_id,
      }
      model_ids.append(model_full_id)

    return {'models': models_list, 'modelIds': model_ids}

  def search_projects(self, search_key, project_id):
    scope = types.SearchCatalogRequest.Scope()
    scope.include_project_ids.append(project_id)
    results = self._datacatalog_client.search_catalog(
        scope=scope,
        query='name:{} projectid:{}'.format(search_key, project_id))

    fetched_results = []
    for result in results:
      resource = result.linked_resource
      result_type = format(result.search_result_subtype)
      if result_type == 'entry.dataset':
        res = re.search('datasets/(.*)', resource)
        dataset = res.group(1)
        fetched_results.append({
            'type': 'dataset',
            'parent': project_id,
            'name': dataset,
            'id': '{}.{}'.format(project_id, dataset)
        })
      elif result_type == 'entry.table':
        res = re.search('datasets/(.*)/tables/(.*)', resource)
        dataset = res.group(1)
        table = res.group(2)
        fetched_results.append({
            'type': 'table',
            'parent': dataset,
            'name': table,
            'id': '{}.{}.{}'.format(project_id, dataset, table)
        })
      elif result_type == 'entry.table.view':
        res = re.search('datasets/(.*)/tables/(.*)', resource)
        dataset = res.group(1)
        view = res.group(2)
        fetched_results.append({
            'type': 'view',
            'parent': dataset,
            'name': view,
            'id': '{}.{}.{}'.format(project_id, dataset, view)
        })
      elif result_type == 'entry.model':
        res = re.search('datasets/(.*)/models/(.*)', resource)
        dataset = res.group(1)
        model = res.group(2)
        fetched_results.append({
            'type': 'model',
            'parent': dataset,
            'name': model,
            'id': '{}.{}.{}'.format(project_id, dataset, model)
        })

    return {'results': fetched_results}

  def create_custom_client(self, project_id):
    return bigquery.Client(project=project_id)

  def get_project(self, custom_client):
    project = custom_client.project
    formatted_project = {
        'id': format(project),
        'name': format(project),
    }
    return formatted_project
