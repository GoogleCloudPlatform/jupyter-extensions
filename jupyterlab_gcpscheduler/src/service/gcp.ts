/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* eslint-disable @typescript-eslint/camelcase */
import { ServerConnection } from '@jupyterlab/services';

import {
  CLOUD_FUNCTION_NAME,
  CLOUD_FUNCTION_REGION,
  IMPORT_DIRECTORY,
} from '../data';
import { ProjectStateService } from './project_state';
import {
  handleApiError,
  TransportService,
  GET,
  POST,
} from 'gcp_jupyterlab_shared';
import { IDocumentManager } from '@jupyterlab/docmanager';
import { Notebook } from '@jupyterlab/notebook';

const SERVICE_ACCOUNT_DOMAIN = 'appspot.gserviceaccount.com';
const IMMEDIATE_JOB_INDICATOR = 'jupyterlab_immediate_notebook';
const SCHEDULED_JOB_INDICATOR = 'jupyterlab_scheduled_notebook';

/**
 * Cloud Scheduler Job
 * https://cloud.google.com/scheduler/docs/reference/rest/v1/projects.locations.jobs#Job
 */
export interface CloudSchedulerJob {
  name: string;
  description: string;
  schedule: string;
  timeZone: string;
  httpTarget: {
    body: string;
    headers: { [name: string]: string };
    httpMethod: string;
    uri: string;
    oidcToken: { serviceAccountEmail: string };
  };
}

/** Message type describing an AI Platform training Job */
export interface RunNotebookRequest {
  imageUri: string;
  inputNotebookGcsPath: string;
  jobId: string;
  masterType: string;
  outputNotebookGcsPath: string;
  scaleTier: string;
  region: string;
  acceleratorType: string;
  acceleratorCount: string;
}

/** List of Jobs returned from AI Platform. */
export type ListAiPlatformJobsResponse = gapi.client.ml.GoogleCloudMlV1__ListJobsResponse;

/** AI Platform Job. */
export type AiPlatformJob = gapi.client.ml.GoogleCloudMlV1__Job;

type StorageObject = gapi.client.storage.Object;

const AI_PLATFORM = 'https://ml.googleapis.com/v1';
const CLOUD_SCHEDULER = 'https://cloudscheduler.googleapis.com/v1';
const STORAGE_DOWNLOAD = 'https://storage.googleapis.com/storage/v1';
const STORAGE_UPLOAD = 'https://storage.googleapis.com/upload/storage/v1';
/**
 * Class to interact with GCP services.
 */
export class GcpService {
  private readonly serverSettings = ServerConnection.defaultSettings;
  private readonly runtimeUrl = `${this.serverSettings.baseUrl}gcp/v1/runtime`;

  constructor(
    private _transportService: TransportService,
    private _projectStateService: ProjectStateService,
    private _documentManager: IDocumentManager
  ) {}

  get projectId() {
    return this._projectStateService.projectId;
  }

  set transportService(transportService: TransportService) {
    this._transportService = transportService;
  }

  /**
   * Uploads the specified Notebook JSON representation to the specified path.
   */
  async uploadNotebook(
    notebookContents: string,
    gcsPath: string
  ): Promise<StorageObject> {
    try {
      const { bucket, object } = this._getGcsPathParts(gcsPath);
      const response = await this._transportService.submit<StorageObject>({
        path: `${STORAGE_UPLOAD}/b/${bucket}/o`,
        method: POST,
        headers: { 'Content-Type': 'application/json' },
        body: notebookContents,
        params: {
          name: object,
          uploadType: 'media',
        },
      });
      return response.result;
    } catch (err) {
      console.error(`Unable to upload Notebook contents to ${gcsPath}`);
      handleApiError(err);
    }
  }

  /**
   * Downloads the notebook located at the gcsPath provided and retrieves the file name from the
   * JSON. Creates a new import directory and attempts to rename it, if it fails then creates
   * the notebook in the already-existing directory and deletes the new directory.
   *
   * @param gcsPath Path to the notebook we want to create.
   */
  async importNotebook(gcsPath: string): Promise<Notebook> {
    const contents = await this.downloadNotebook(gcsPath);

    const retrievedName = JSON.parse(contents)
      .metadata?.papermill?.input_path?.split('/')
      .slice(-1)[0];
    const fileName = retrievedName ? retrievedName : 'untitled.ipynb';

    const newDirectory = await this._documentManager.newUntitled({
      type: 'directory',
    });

    try {
      await this._documentManager.rename(newDirectory.path, IMPORT_DIRECTORY);
    } catch (err) {
      // Expect a 409 error if directory exists. Otherwise throw the error
      if (err.message.indexOf('409') === -1) {
        throw err;
      }
      this._documentManager.deleteFile(newDirectory.path);
    }

    return this.createNotebook(fileName, contents);
  }

  private createNotebook(fileName: string, contents: string): Notebook {
    const widget = this._documentManager.createNew(
      IMPORT_DIRECTORY + fileName,
      'Notebook'
    ) as Notebook;
    widget.model.fromString(contents);
    return widget;
  }

  async downloadNotebook(gcsPath: string): Promise<string> {
    try {
      const { bucket, object } = this._getGcsPathParts(gcsPath);
      const response = await this._transportService.submit<string>({
        path: `${STORAGE_DOWNLOAD}/b/${bucket}/o/${encodeURIComponent(object)}`,
        params: {
          alt: 'media',
        },
      });
      return typeof response.result === 'string'
        ? response.result
        : JSON.stringify(response.result);
    } catch (err) {
      console.error(`Unable to download Notebook contents at ${gcsPath}`);
      handleApiError(err);
    }
  }

  /**
   * Submits a Notebook for recurring scheduled execution on AI Platform via a
   * new Cloud Scheduler job.
   * @param request
   * @param regionName
   * @param serviceAccountEmail
   * @param schedule
   */
  async scheduleNotebook(
    request: RunNotebookRequest,
    regionName: string,
    schedule: string
  ): Promise<CloudSchedulerJob> {
    let timeZone = 'America/New_York';
    try {
      timeZone = Intl.DateTimeFormat().resolvedOptions().timeZone;
    } catch (err) {
      console.warn('Unable to determine timezone');
    }

    try {
      const projectId = await this.projectId;
      const locationPrefix = `projects/${projectId}/locations/${regionName}/jobs`;
      const requestBody: CloudSchedulerJob = {
        name: `${locationPrefix}/${request.jobId}`,
        description: SCHEDULED_JOB_INDICATOR,
        schedule,
        timeZone,
        httpTarget: {
          body: btoa(
            JSON.stringify(this._buildAiPlatformJobRequest(request, true))
          ),
          headers: { 'Content-Type': 'application/json' },
          httpMethod: POST,
          oidcToken: {
            serviceAccountEmail: `${projectId}@${SERVICE_ACCOUNT_DOMAIN}`,
          },
          uri: `https://${CLOUD_FUNCTION_REGION}-${projectId}.cloudfunctions.net/${CLOUD_FUNCTION_NAME}`,
        },
      };
      const response = await this._transportService.submit<CloudSchedulerJob>({
        path: `${CLOUD_SCHEDULER}/${locationPrefix}`,
        method: POST,
        body: requestBody,
      });
      return response.result;
    } catch (err) {
      console.error('Unable to create Cloud Scheduler job');
      handleApiError(err);
    }
  }

  /**
   * Submits a Notebook for immediate execution on AI Platform.
   * @param cloudFunctionUrl
   * @param request
   */
  async runNotebook(request: RunNotebookRequest): Promise<AiPlatformJob> {
    try {
      const projectId = await this.projectId;
      const response = await this._transportService.submit<AiPlatformJob>({
        path: `${AI_PLATFORM}/projects/${projectId}/jobs`,
        method: POST,
        body: this._buildAiPlatformJobRequest(request),
      });
      return response.result;
    } catch (err) {
      console.error('Unable to submit Notebook to AI Platform');
      handleApiError(err);
    }
  }

  /**
   * Gets list of AiPlatformJobs
   * @param cloudFunctionUrl
   * @param request
   */
  async listNotebookJobs(
    pageToken?: string
  ): Promise<ListAiPlatformJobsResponse> {
    try {
      const filter = [SCHEDULED_JOB_INDICATOR, IMMEDIATE_JOB_INDICATOR]
        .map(v => `labels.job_type=${v}`)
        .join(' OR ');
      const projectId = await this.projectId;
      const params: { [k: string]: string } = { filter };
      if (pageToken) {
        params['pageToken'] = pageToken;
      }
      const response = await this._transportService.submit<
        ListAiPlatformJobsResponse
      >({
        path: `${AI_PLATFORM}/projects/${projectId}/jobs`,
        params,
      });
      return response.result;
    } catch (err) {
      console.error('Unable to list AI Platform Notebook jobs');
      handleApiError(err);
    }
  }

  async getImageUri(): Promise<string> {
    // TODO need to check if image exist
    const imageUriPrefix = 'gcr.io/deeplearning-platform-release/';
    const runtimeEnv = await this._getRuntimeEnv();
    if (!runtimeEnv) return '';

    const lastDotIndex = runtimeEnv.lastIndexOf('.');
    return `${imageUriPrefix}${runtimeEnv.substr(0, lastDotIndex)}`;
  }

  private async _getRuntimeEnv(): Promise<string> {
    try {
      const response = await ServerConnection.makeRequest(
        this.runtimeUrl,
        { method: GET },
        this.serverSettings
      );
      return await response.text();
    } catch (err) {
      console.warn('Unable to obtain runtime environment');
      return '';
    }
  }

  private _buildAiPlatformJobRequest(
    request: RunNotebookRequest,
    isScheduled = false
  ): AiPlatformJob {
    return {
      jobId: request.jobId,
      labels: {
        job_type: isScheduled
          ? SCHEDULED_JOB_INDICATOR
          : IMMEDIATE_JOB_INDICATOR,
      },
      trainingInput: {
        args: [
          'nbexecutor',
          '--input-notebook',
          request.inputNotebookGcsPath,
          '--output-notebook',
          request.outputNotebookGcsPath,
        ],
        masterConfig: {
          imageUri: request.imageUri,
          acceleratorConfig: {
            count: request.acceleratorCount || undefined,
            type: request.acceleratorType || undefined,
          },
        },
        masterType: request.masterType || undefined,
        region: request.region,
        scaleTier: request.scaleTier,
      },
    } as AiPlatformJob;
  }

  private _getGcsPathParts(
    gcsPath: string
  ): { bucket: string; object: string } {
    if (gcsPath.startsWith('gs://')) {
      gcsPath = gcsPath.slice(5);
    }
    const pathParts = gcsPath.split('/');
    return {
      bucket: pathParts[0],
      object: pathParts.slice(1).join('/'),
    };
  }
}
