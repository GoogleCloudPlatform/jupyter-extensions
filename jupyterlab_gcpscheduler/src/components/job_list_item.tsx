/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { withStyles } from '@material-ui/core';
import { MoreVert } from '@material-ui/icons';
import * as csstips from 'csstips';
import {
  css,
  IconButtonMenu,
  SmallMenuItem,
  GrayPending,
  GreenCheck,
  RedClose,
} from 'gcp_jupyterlab_shared';
import * as React from 'react';
import { classes, stylesheet } from 'typestyle';

import { AI_PLATFORM_LINK } from '../data';
import { AiPlatformJob, GcpService } from '../service/gcp';

interface Props {
  job: AiPlatformJob;
  projectId: string;
  gcpService: GcpService;
}

const localStyles = stylesheet({
  item: {
    alignItems: 'center',
    borderBottom: 'var(--jp-border-width) solid var(--jp-border-color2)',
    listStyle: 'none',
    height: '40px',
    paddingRight: '8px',
    $nest: {
      '&:hover': {
        backgroundColor: 'var(--jp-layout-color2)',
      },
    },
    ...csstips.horizontal,
  },
  details: {
    paddingLeft: '4px',
    ...csstips.flex,
  },
  jobLink: {
    textDecoration: 'none',
  },
  jobTime: {
    color: 'var(--jp-content-font-color2)',
    fontSize: '9px',
    textAlign: 'right',
    ...csstips.flex,
  },
  menuLink: {
    display: 'block',
    height: '100%',
    width: '100%',
  },
});

const MenuIcon = withStyles({
  root: {
    fontSize: '20px',
  },
})(MoreVert);

const VIEWER_LINK_BASE = 'https://notebooks.cloud.google.com/view';
const DOWNLOAD_LINK_BASE = 'https://storage.cloud.google.com';
const SUCCEEDED = 'SUCCEEDED';

function getIconForJobState(state: string): JSX.Element {
  if (state === 'SUCCEEDED') {
    return <GreenCheck />;
  } else if (state === 'FAILED') {
    return <RedClose />;
  }
  return <GrayPending />;
}

/** Notebook job list item */
export function JobListItem(props: Props) {
  const { gcpService, job, projectId } = props;
  const endTime = new Date(job.endTime || job.createTime);

  const gcsFile =
    job.trainingInput &&
    job.trainingInput.args &&
    job.trainingInput.args[4].slice(5);

  /*
    Extract the bucket and jobName from the GCS URI. The actual AI Platform job
    ID may have an additional timestamp if it was submitted as a recurring run
  */
  const [bucket, jobName, ...object] = gcsFile.split('/');
  const name = jobName.replace('_', ' ');
  const encodedObjectPath = [jobName, ...object]
    .map(p => encodeURIComponent(p))
    .join('/');
  const jobLink = `${AI_PLATFORM_LINK}/${job.jobId}?project=${projectId}`;
  const viewerLink = `${VIEWER_LINK_BASE}/${bucket}/${encodedObjectPath}`;
  const downloadLink = `${DOWNLOAD_LINK_BASE}/${gcsFile}`;
  return (
    <li className={localStyles.item}>
      <div title={`Status: ${job.state}`}>{getIconForJobState(job.state)}</div>
      <div className={localStyles.details}>
        <div>
          <a
            className={classes(css.link, localStyles.jobLink)}
            href={jobLink}
            target="_blank"
            title="View AI Platform Job"
          >
            {name}
          </a>
        </div>
        <div>
          <span className={localStyles.jobTime}>
            {endTime.toLocaleString()}
          </span>
        </div>
      </div>
      <div>
        {job.state === SUCCEEDED && (
          <IconButtonMenu
            icon={<MenuIcon />}
            menuItems={menuCloseHandler => [
              <SmallMenuItem key="viewNotebook">
                <a
                  className={localStyles.menuLink}
                  href={viewerLink}
                  target="_blank"
                  title="View notebook output in HTML format"
                  onClick={menuCloseHandler}
                >
                  View
                </a>
              </SmallMenuItem>,
              <SmallMenuItem
                id="import"
                key="importNotebook"
                onClick={() => gcpService.importNotebook(gcsFile)}
              >
                Import
              </SmallMenuItem>,
              <SmallMenuItem key="getNotebookPath" onClick={menuCloseHandler}>
                <a
                  className={localStyles.menuLink}
                  href={downloadLink}
                  target="_blank"
                  title="Download the notebook output from Google Cloud Storage"
                  onClick={menuCloseHandler}
                >
                  Download
                </a>
              </SmallMenuItem>,
            ]}
          />
        )}
      </div>
    </li>
  );
}
