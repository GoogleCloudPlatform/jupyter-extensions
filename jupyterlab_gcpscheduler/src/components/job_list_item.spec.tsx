/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { shallow } from 'enzyme';
import * as React from 'react';

import { JobListItem } from './job_list_item';
import { TEST_PROJECT, getAiPlatformJob } from '../test_helpers';
import { GcpService } from '../service/gcp';
import { IconButtonMenu } from 'gcp_jupyterlab_shared';

const toLocaleString = Date.prototype.toLocaleString;

describe('JobListItem', () => {
  const mockImportNotebook = jest.fn();
  const mockGcpService = ({
    importNotebook: mockImportNotebook,
  } as unknown) as GcpService;

  beforeEach(() => {
    Date.prototype.toLocaleString = jest
      .fn()
      .mockReturnValue('5/1/2020, 3:09:42 PM');
  });

  afterEach(() => {
    Date.prototype.toLocaleString = toLocaleString;
  });

  it('Calls GcpService on Import', () => {
    const jobListItem = shallow(
      <JobListItem
        gcpService={mockGcpService}
        projectId={TEST_PROJECT}
        job={getAiPlatformJob()}
      />
    );

    jobListItem
      .find(IconButtonMenu)
      .dive()
      .findWhere(w => w.text() === 'Import')
      .first()
      .simulate('click');

    expect(mockGcpService.importNotebook).toHaveBeenCalledWith(
      'test-project/notebook_job1/job1.ipynb'
    );
  });

  it('Renders for successful job', () => {
    const component = shallow(
      <JobListItem
        gcpService={mockGcpService}
        projectId={TEST_PROJECT}
        job={getAiPlatformJob()}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it('Renders for failed job', () => {
    const job = getAiPlatformJob();
    job.state = 'FAILED';
    const component = shallow(
      <JobListItem
        gcpService={mockGcpService}
        projectId={TEST_PROJECT}
        job={getAiPlatformJob()}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it('Renders for running job', () => {
    const job = getAiPlatformJob();
    job.state = 'RUNNING';
    const component = shallow(
      <JobListItem
        gcpService={mockGcpService}
        projectId={TEST_PROJECT}
        job={getAiPlatformJob()}
      />
    );
    expect(component).toMatchSnapshot();
  });
});
