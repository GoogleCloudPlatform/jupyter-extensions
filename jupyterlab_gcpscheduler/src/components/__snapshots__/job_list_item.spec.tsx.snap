// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobListItem Renders for failed job 1`] = `
<li
  className="item_f7pj24d"
>
  <div
    title="Status: SUCCEEDED"
  >
    <WithStyles(CheckIcon) />
  </div>
  <div
    className="details_f1grbse3"
  >
    <div>
      <a
        className="link_fg3laeu jobLink_f2antnr"
        href="https://console.cloud.google.com/ai-platform/jobs/notebook_job1_abcxyz?project=test-project"
        target="_blank"
        title="View AI Platform Job"
      >
        notebook job1
      </a>
    </div>
    <div>
      <span
        className="jobTime_fueeqh8"
      >
        5/1/2020, 3:09:42 PM
      </span>
    </div>
  </div>
  <div>
    <IconButtonMenu
      icon={<WithStyles(MoreVertIcon) />}
      menuItems={[Function]}
    />
  </div>
</li>
`;

exports[`JobListItem Renders for running job 1`] = `
<li
  className="item_f7pj24d"
>
  <div
    title="Status: SUCCEEDED"
  >
    <WithStyles(CheckIcon) />
  </div>
  <div
    className="details_f1grbse3"
  >
    <div>
      <a
        className="link_fg3laeu jobLink_f2antnr"
        href="https://console.cloud.google.com/ai-platform/jobs/notebook_job1_abcxyz?project=test-project"
        target="_blank"
        title="View AI Platform Job"
      >
        notebook job1
      </a>
    </div>
    <div>
      <span
        className="jobTime_fueeqh8"
      >
        5/1/2020, 3:09:42 PM
      </span>
    </div>
  </div>
  <div>
    <IconButtonMenu
      icon={<WithStyles(MoreVertIcon) />}
      menuItems={[Function]}
    />
  </div>
</li>
`;

exports[`JobListItem Renders for successful job 1`] = `
<li
  className="item_f7pj24d"
>
  <div
    title="Status: SUCCEEDED"
  >
    <WithStyles(CheckIcon) />
  </div>
  <div
    className="details_f1grbse3"
  >
    <div>
      <a
        className="link_fg3laeu jobLink_f2antnr"
        href="https://console.cloud.google.com/ai-platform/jobs/notebook_job1_abcxyz?project=test-project"
        target="_blank"
        title="View AI Platform Job"
      >
        notebook job1
      </a>
    </div>
    <div>
      <span
        className="jobTime_fueeqh8"
      >
        5/1/2020, 3:09:42 PM
      </span>
    </div>
  </div>
  <div>
    <IconButtonMenu
      icon={<WithStyles(MoreVertIcon) />}
      menuItems={[Function]}
    />
  </div>
</li>
`;
